File: packages/typescript/examples/quickstart.ts
/**
 * Create an agent, start a run, follow it live, then page through history.
 *
 *   UARP_API_KEY=uarp_... node --experimental-strip-types examples/quickstart.ts
 */
import {
  APIError,
  RateLimitError,
  UarpClient,
  UnprocessableEntityError,
  type Agent,
} from '../src/index.js';

const client = new UarpClient({
  apiKey: process.env.UARP_API_KEY,
  // Everything below is optional; these are the defaults.
  timeout: 60_000,
  maxRetries: 2,
});

async function createAgent(): Promise<Agent> {
  return client.agents.create({
    name: 'quickstart',
    model: {
      provider: 'openai_compat',
      model_ref: 'gpt-4o-mini',
      capabilities: {},
    },
    prompts: { system: 'You are a concise assistant.' },
  });
}

async function runAndFollow(agentId: string): Promise<void> {
  const run = await client.runs.create({
    agent_id: agentId,
    input: { message: 'Summarise the last deploy.' },
  });
  const runId = run.run_id;

  // The stream reconnects with Last-Event-ID if the connection drops.
  for await (const event of client.runs.streamRunEvents(runId)) {
    if (event.event === 'llm.chunk') {
      process.stdout.write(event.json<{ text?: string }>().text ?? '');
    }
    if (event.event === 'run.completed' || event.event === 'run.failed') break;
  }
}

async function listEverything(): Promise<void> {
  // `listAll` walks every page; `list` returns one page plus its cursor.
  for await (const agent of client.agents.listAll({ limit: 50 })) {
    console.log(`${agent.agent_id}  ${agent.name}`);
  }
}

async function main(): Promise<void> {
  try {
    const agent = await createAgent();
    await runAndFollow(agent.agent_id);
    await listEverything();
  } catch (error) {
    if (error instanceof UnprocessableEntityError) {
      for (const failure of error.validationErrors) {
        console.error(`invalid ${failure.field}: ${failure.message}`);
      }
    } else if (error instanceof RateLimitError) {
      console.error(`rate limited; retry after ${error.retryAfterSeconds ?? '?'}s`);
    } else if (error instanceof APIError) {
      console.error(`${error.status} ${error.message} (${error.correlationId ?? 'no id'})`);
    } else {
      throw error;
    }
  }
}

await main();

File: packages/rust/examples/quickstart.rs
//! Create an agent, start a run, follow it live, then page through history.
//!
//! ```sh
//! UARP_API_KEY=uarp_... cargo run --example quickstart
//! ```

use std::pin::pin;

use futures_util::StreamExt;
use uarp_sdk::api::ListAgentsParams;
use uarp_sdk::models::{Agent, CreateAgentRequest, CreateRunRequest};
use uarp_sdk::{ApiErrorKind, Client, Error};

async fn create_agent(client: &Client) -> Result<Agent, Error> {
    let request = CreateAgentRequest {
        name: "quickstart".into(),
        ..Default::default()
    };
    client.agents().create(&request).await
}

async fn run_and_follow(client: &Client, agent_id: &str) -> Result<(), Error> {
    let run = client
        .runs()
        .create(&CreateRunRequest {
            agent_id: agent_id.to_string(),
            ..Default::default()
        })
        .await?;

    let runs = client.runs();
    let params = Default::default();
    // The stream reconnects with Last-Event-ID if the connection drops.
    let mut events = pin!(runs.stream_run_events(&run.run_id, &params));
    while let Some(event) = events.next().await {
        let event = event?;
        match event.event.as_str() {
            "llm.chunk" => print!("{}", event.data),
            "run.completed" | "run.failed" => break,
            _ => {}
        }
    }
    Ok(())
}

async fn list_everything(client: &Client) -> Result<(), Error> {
    let agents = client.agents();
    let params = ListAgentsParams {
        limit: Some(50),
        ..Default::default()
    };

    // `list_all` walks every page; `list` returns one page plus its cursor.
    let mut stream = pin!(agents.list_all(&params));
    while let Some(agent) = stream.next().await {
        let agent = agent?;
        println!("{}  {}", agent.agent_id, agent.name);
    }
    Ok(())
}

#[tokio::main]
async fn main() {
    let client = match Client::from_env() {
        Ok(client) => client,
        Err(error) => {
            eprintln!("{error}");
            std::process::exit(1);
        }
    };

    if let Err(error) = async {
        let agent = create_agent(&client).await?;
        run_and_follow(&client, &agent.agent_id).await?;
        list_everything(&client).await
    }
    .await
    {
        match &error {
            Error::Api(api) if api.kind() == ApiErrorKind::UnprocessableEntity => {
                for failure in &api.problem.errors {
                    eprintln!(
                        "invalid {}: {}",
                        failure.field.as_deref().unwrap_or("?"),
                        failure.message.as_deref().unwrap_or("?")
                    );
                }
            }
            Error::Api(api) if api.kind() == ApiErrorKind::RateLimit => {
                eprintln!("rate limited; retry after {:?}s", api.retry_after_seconds());
            }
            other => eprintln!("{other}"),
        }
        std::process::exit(1);
    }
}
